torch/csrc/jit/passes/canonicalize.cpp

```diff
@@ -129,20 +129,33 @@ bool isBefore(const Use& a, const Use& b) {
   return isBefore(a.user, b.user);
 }
 
-std::vector<c10::optional<Use>> gatherFirstUses(at::ArrayRef<Value*> values) {
-  return fmap(values, [](Value* v) -> c10::optional<Use> {
-    if (v->uses().size() == 0) {
-      return c10::nullopt;
-    }
-    Use first_use = v->uses()[0];
-    for (size_t i = 1; i < v->uses().size(); ++i) {
-      auto n_use = v->uses()[i];
-      if (!isBefore(first_use, n_use)) {
-        first_use = n_use;
-      }
+bool isAfter(const Use& a, const Use& b) {
+  return !isBefore(a, b);
+}
+
+bool isBeforeOrAfter(const Use& a, const Use& b, bool checking_before) {
+  return checking_before ? isBefore(a, b) : isAfter(a, b);
+}
+
+c10::optional<const Use> firstOrLastUse(Value* v, bool find_first) {
+  if (v->uses().size() == 0) {
+    return c10::nullopt;
+  }
+  Use extreme_use = v->uses()[0];
+  for (size_t i = 1; i < v->uses().size(); ++i) {
+    auto n_use = v->uses()[i];
+    if (!isBeforeOrAfter(extreme_use, n_use, find_first)) {
```

Comment:
nitpick: i don't if there's a lot of value in refactoring that ternary into a helper given there's just one use of it. it seems to make code a little bit harder to follow